docs/source/developer/extension_migration.rst

```diff
@@ -0,0 +1,162 @@
+  - ``p-`` DOM events such as ``p-dragenter`` should be updated to ``lm-``,
+    e.g. ``lm-dragenter``
+
+Updating former ``@jupyterlab/coreutils`` imports
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+JupyterLab 2.0 introduces several new packages with classes and tokens that
+have been moved out of ``@jupyterlab/coreutils`` into their own packages. These
+exports have been moved:
+
+============================  =================================
+ Export                        Package
+============================  =================================
+ ``DataConnector``             ``@jupyterlab/statedb``
+ ``Debouncer``                 ``@lumino/polling``
+ ``DefaultSchemaValidator``    ``@jupyterlab/settingregistry``
+ ``IDataConnector``            ``@jupyterlab/statedb``
+ ``IObjectPool``               ``@jupyterlab/statedb``
+ ``IPoll``                     ``@lumino/polling``
+ ``IRateLimiter``              ``@lumino/polling``
+ ``IRestorable``               ``@jupyterlab/statedb``
+ ``IRestorer``                 ``@jupyterlab/statedb``
+ ``ISchemaValidator``          ``@jupyterlab/settingregistry``
+ ``ISettingRegistry``          ``@jupyterlab/settingregistry``
+ ``IStateDB``                  ``@jupyterlab/statedb``
+ ``nbformat``                  ``@jupyterlab/nbformat``
+ ``Poll``                      ``@lumino/polling``
+ ``RateLimiter``               ``@lumino/polling``
+ ``RestorablePool``            ``@jupyterlab/statedb``
+ ``SettingRegistry``           ``@jupyterlab/settingregistry``
+ ``Settings``                  ``@jupyterlab/settingregistry``
+ ``StateDB``                   ``@jupyterlab/statedb``
+ ``Throttler``                 ``@lumino/polling``
+============================  =================================
+
+.. tip::
+  It might be helpful to delete ``node_modules`` and ``yarn.lock`` when
+  updating these libraries.
+
+Using ``Session`` and ``SessionContext`` to manage kernel sessions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+``ConsolePanel`` and ``NotebookPanel`` now expose a
+``sessionContext: ISessionContext`` attribute that allows for a uniform way to
+interact with kernel sessions.
+
+Any widget that matches the ``interface IDocumentWidget`` has a
+``context: DocumentRegistry.IContext`` attribute with a
+``sessionContext: ISessionContext`` attribute.
+
+For example, consider how the ``@jupyterlab/debugger`` extension's
+``DebuggerService`` updated its ``isAvailable()`` method.
+
+.. figure:: extension_migration_session.png
+   :align: center
+   :class: jp-screenshot
+   :alt: Updating the isAvailable method of the debugger service
+
+   From the `PR migrating the debugger extension to JupyterLab 2.0 <https://github.com/jupyterlab/debugger/pull/337/files#diff-22ccf3ebb0cb6b300ee90a38b88edff8>`__
+
+.. note::
+
+  "One notable change: ``await client.kernel.ready`` becomes
+  ``await sessionContext.session.kernel.info`` to check when a kernel is
+  ready." --
```

Comment:
Thanks for adding this!

We might want to refine this note a little bit. After looking into it more (and following up on a discussion on gitter with Jason), a kernel should be ready to use if it exists. Messages will be queued in case the websocket connection hasn't been established yet.

For the debugger we wait for kernel info to be able to look up the `debugger` key.

Also we might want to recommend using the optional chaining syntax (similar to how it is used across the code base): 

```typescript
await sessionContext.session?.kernel?.info
```